### PR TITLE
Circe

### PR DESCRIPTION
**Pull request by mdedetrich**
This is the commit which changes from play-json to circe. Fixes https://github.com/mdedetrich/stripe-scala/issues/15 and also fixes some models which were incorrect + some typos

One thing to note is that this uses streaming-json which has some advantages (can handle large JSON payloads much better in terms of memory throughput/backpressure) however json streams can only be materialized once which means we don't get accumulated errors. Note that this is nothing specific to Circe, its how the JSON streaming library works.

Also in general, formatting has been cleaned up in a lot of areas and a lot of typos have also been fixed (also some incorrect fields have been fixed up too)

`readNullableOrEmptyJsObject ` has also been removed, I think this obfuscated things too much and it didn't seem to bring many advantages

### CODE REVIEW

File: src/main/scala/org/mdedetrich/stripe/v1/Events.scala [hunk 1]
```diff
@@ -19,194 +19,87 @@ object Events extends LazyLogging {
...
-    }
-
-    override def writes(o: StripeObject): JsValue = Json.obj()
+    case object ChargeCaptured                    extends Type("charge.captured")
+    case object ChargeFailed                      extends Type("charge.failed")
+    case object ChargeRefunded                    extends Type("charge.refunded")
+    case object ChargeSucceeded                   extends Type("charge.succeeded")
+    case object ChargeUpdated                     extends Type("charge.updated")
+    case object ChargeDisputeClosed               extends Type("charge.dispute.closed")
+    case object ChargeDisputeCreated              extends Type("harge.dispute.created")
```
Comment (leonardehrenfried) [inline src/main/scala/org/mdedetrich/stripe/v1/Events.scala]: There is a typo here that has come from master: a 'c' is missing in charge.

**Comment by leonardehrenfried** [comment]
I like this change a lot: well done!

I will be taking some time tomorrow for a more thorough review but on first glance it looks good.

File: src/main/scala/org/mdedetrich/stripe/v1/Charges.scala [hunk 2]
```diff
@@ -36,44 +35,55 @@ object Charges extends LazyLogging {
...
       override val entryName = id
     }
 
-    object StripeReport extends Enum[StripeReport] with PlayJsonEnum[StripeReport] {
+    object StripeReport extends Enum[StripeReport] {
       val values = findValues
+
       case object Fraudulent extends StripeReport("fraudulent")
+
+      implicit val stripeReportDecoder: Decoder[StripeReport] = enumeratum.Circe.decoder(StripeReport)
```
Comment (leonardehrenfried) [inline src/main/scala/org/mdedetrich/stripe/v1/Charges.scala:57]: Isn't there a circe support for enumeratum?

File: src/main/scala/org/mdedetrich/stripe/v1/Charges.scala [hunk 3]
```diff
@@ -315,99 +391,210 @@ object Charges extends LazyLogging {
...
+    "invoice",
+    "livemode",
+    "metadata",
+    "order",
+    "paid",
+    "receipt_email",
+    "receipt_number"
+  )(
+    Tuple22.apply(
+      _: String,
```
Comment (leonardehrenfried) [inline src/main/scala/org/mdedetrich/stripe/v1/Charges.scala:419]: Why do you habe to declare the types here and not in the other decoders?

File: src/main/scala/org/mdedetrich/stripe/v1/Events.scala (hunk 1 above)
Comment (mdedetrich) [reply src/main/scala/org/mdedetrich/stripe/v1/Events.scala]: Fixed, thanks for the pickup!

File: src/main/scala/org/mdedetrich/stripe/v1/Events.scala [hunk 4]
```diff
@@ -19,194 +19,87 @@ object Events extends LazyLogging {
...
-
-    case object Ping extends Type("ping")
-  }
-
-  private implicit val stripeObjectReads = new Format[StripeObject] {
-
-    private val error = JsError("error.expected.stripeObject")
-
-    private def parseStripeObject(name: String, json: JsObject): JsResult[StripeObject] = name match {
-      case "customer"        => json.validate[Customers.Customer]
```
Comment (leonardehrenfried) [inline src/main/scala/org/mdedetrich/stripe/v1/Events.scala:170]: In circe, how is it decided which version decoder is to be used?

File: src/test/scala/org/mdedetrich/stripe/v1/EventsSpec.scala [hunk 5]
```diff
@@ -28,26 +33,34 @@ class EventsSpec extends WordSpec with Matchers {
       s"parse accounts.update from $filename" in {
         val in = this.getClass.getResourceAsStream(s"/events/$filename")
         in should not be null
-        val json                = Json.parse(in)
-        val JsSuccess(event, _) = json.validate[Event]
+
+        val string = scala.io.Source.fromInputStream(in).mkString
```
Comment (leonardehrenfried) [inline src/test/scala/org/mdedetrich/stripe/v1/EventsSpec.scala:37]: I think in future we should be using utility methods for this. This shouldn't block the merge though.

File: build.sbt [hunk 6]
```diff
@@ -30,22 +31,24 @@ Defaults.itSettings
...
+  "de.knutwalker"              %% "akka-stream-circe" % akkaStreamJson,
+  "de.knutwalker"              %% "akka-http-circe"   % akkaStreamJson,
+  "io.circe"                   %% "circe-core"        % circeVersion,
+  "io.circe"                   %% "circe-generic"     % circeVersion,
+  "io.circe"                   %% "circe-parser"      % circeVersion,
+  "com.beachape"               %% "enumeratum"        % enumeratumVersion,
+  "com.beachape"               %% "enumeratum-circe"  % enumeratumCirceVersion,
+  "com.iheart"                 %% "ficus"             % "1.3.4",
+  "com.typesafe.scala-logging" %% "scala-logging"     % "3.4.0",
+  "com.netaporter"             %% "scala-uri"         % "0.4.13" exclude ("io.spray", "spray-json_2.11"),
```
Comment (leonardehrenfried) [inline build.sbt:49]: Didn't you mention that you'd like to get right of scala-uri?

**Comment by leonardehrenfried** [comment]
Okay, I managed to review this earlier than I though. It looks like solid work and the test are passing as well.

I think this can be merged.

Now that play-json and dispatch are replaced, are we good to release 0.2.0?

**Review by leonardehrenfried — APPROVED** [review]

**Comment by mdedetrich** [comment]
> Now that play-json and dispatch are replaced, are we good to release 0.2.0?

There are a couple more cleanups that I want to do, like #24 and #20  and possibly #13

**Pull request merged**